TCD1304.py

```diff
@@ -0,0 +1,80 @@
+'''
+Requests for the Code:
+the code should have an power Output on the Voltagesource VDD=5V /VAD=5V
+the code should build a Masterclock with a frequency inbetween 0.8 to 4 MHz, typical 2 MHz. The Voltagerange of the Clock 
+shall be 0-4V
+The ICG is a squarewave with a high level of typical 4V over the time of 3694 Elements of the Clock.
+The SH is a squarewave with a high level of typical 4V and indicates the integrationtime. 
+The integrationtime is started with the falling flank of the signal and and lasts till the next falling flank.
+The analog input of the sensor shall be read by the the osilloscope.
+the analog voltage signal shall be saved over the time and be integrated to get a digital signal.
+the algotithims still needs to be definded.
+
+import argparse
+import csv
+import json
+import platform
+import os.path
+import shutil
+import sys
+import time
+from itertools import zip_longest
+from typing import List, Tuple
+
+import numpy as np
+
+'''
+
+
+import time
+import math
+from pslab.instrument.waveform_generator import PWMGenerator
+from pslab.instrument.oscilloscope import Oscilloscope
+from pslab.serial_handler import SerialHandler
+from pslab.instrument.logic_analyzer import LogicAnalyzer
+from pslab.instrument.power_supply import PowerSupply
+
+frequency_master_clock = fmc
+fmc = float (2e6)
+integration_time = intti
+intti = float (1.85e-3) # time in seconds
+
+class tcd1304():
+    def __init__(
+        self, #steht Standartmäßig bei __init__funktionen an erster Stelle
+        min_masterclock_frequency: int = 0.8e6,
+        max_masterclock_frequency: int = 4e6,
+        frequency: float = 2e6, #Standart Frequenz der Masterclock sind 2 MHz
+        min_volatagerange_clock: int = 0,
+        max_volatagerange_clock: int = 4,
+        ): 
+
+        self.min_masterclock_frequency = min_masterclock_frequency #Hier wird initialisiert, wenn mich nicht alles täsucht
+        self.max_masterclock_frequency = max_masterclock_frequency
+        self.frequency = frequency
+        self.min_volatagerange_clock = min_volatagerange_clock
+        self.max_volatagerange_clock = max_volatagerange_clock
+
+    def power_source ():        # puts a voltage of 4V on PV1
+        ps = PowerSupply()
+        ps.pv1 = 4
+        ps.pv1
+    
+    def master_clock ():        # puts PWM with frequency of 2MHz on SQ1
+        pwmgen = PWMGenerator()
+        pwmgen.generate(["SQ1"], fmc, [0.5])       
+    
+    def sh_clock ():            #  on SQ2 sets pulse for the integrationtime
+        ''' 
+        integrationtime is started by the falling flank of voltageoutput. 
+        the time for integration must be longer than 3_694 Elements of fmc.
+        sh clock is started before icg is started. normal state of sh is a low
+        level. high level can be quite short
+        '''
+
+    def icg_clock ():           #  on SQ3 starts and stops the reading of the sensor

```

Comment:
This is the slowest clock, which _may_ be possible to control by calling `PWMGenerator.set_state`:
```python
# integration_elements is the number of output elements, see https://tcd1304.files.wordpress.com/2016/07/tcd1304-timing-es.png
icg_pulses = integration_time / integration_elements
for pulse in range(icg_pulses):
    self.pwmgen.set_state(sq3=True)
    time.sleep(icg_period / 2)
    self.pwmgen.set_state(sq3=False)
    time.sleep(icg_period / 2)
```
It remains to be seen if this method actually works or not. The sensor may require that the icg clock is more in sync with the sh clock than this. If so we will try something else.

With this method the integration time must be fairly large, and/or the number of output elements fairly small. As a rule of thumb, with this method `icg_period` should be no smaller than 1 ms.

Reply:
This is not quite right. I misunderstood the meaning of integration time; I thought it referred to the whole data collection period, but it is actually just the period of the ICG signal. The above code snippet should be:
```python
for pulse in range(integration_elements):
    self.pwmgen.set_state(sq3=True)
    time.sleep(integration_time / 2)
    self.pwmgen.set_state(sq3=False)
    time.sleep(integration_time / 2)
```
Which means that the `integration_time` should be at least 1 ms.